crates/net/downloaders/src/headers/concurrent.rs

```diff
@@ -0,0 +1,719 @@
+    ///
+    /// This will validate all headers and insert them into the validated buffer.
+    ///
+    /// Returns an error if the given headers are invalid
+    #[allow(clippy::result_large_err)]
+    fn process_next_headers(
+        &mut self,
+        request: HeadersRequest,
+        headers: Vec<Header>,
+        peer_id: PeerId,
+        retries: usize,
+    ) -> Result<(), HeadersResponseError> {
+        for parent in headers {
+            let parent = parent.seal();
+
+            // Validate the headers
+            if let Some(header) = self.earliest_header() {
+                if let Err(error) = self.validate(header, &parent) {
+                    return Err(HeadersResponseError {
+                        request,
+                        retries,
+                        peer_id: Some(peer_id),
+                        error,
+                    })
+                }
+            } else if parent.hash() != self.next_chain_tip_hash {
+                return Err(HeadersResponseError {
+                    request,
+                    retries,
+                    peer_id: Some(peer_id),
+                    error: DownloadError::InvalidTip {
+                        received: parent.hash(),
+                        expected: self.next_chain_tip_hash,
+                    },
+                })
+            }
+
+            // update tracked block info
+            self.next_chain_tip_hash = parent.parent_hash;
+            self.next_chain_tip_block_number = parent.number - 1;
+            self.queued_validated_headers.push(parent);
+        }
+
+        Ok(())
+    }
+
+    /// Invoked when we received a response
+    #[allow(clippy::result_large_err)]
+    fn on_headers_outcome(
+        &mut self,
+        response: HeadersRequestOutcome,
+    ) -> Result<(), HeadersResponseError> {
+        let requested_block_number = response.block_number();
+        let HeadersRequestOutcome { request, retries, outcome } = response;
+
+        match outcome {
+            Ok(res) => {
+                let (peer_id, headers) = res.split();
+                let mut headers = headers.0;
+
+                // sort headers from highest to lowest block number
+                headers.sort_unstable_by_key(|h| Reverse(h.number));
+
+                if headers.is_empty() {
+                    return Err(HeadersResponseError {
+                        request,
+                        retries,
+                        peer_id: Some(peer_id),
+                        error: DownloadError::EmptyResponse,
+                    })
+                }
+
+                // validate the response
+                let highest = &headers[0];
+
+                if highest.number != requested_block_number {
+                    // TODO return error, invalid response
+                }
+
+                if (headers.len() as u64) != request.limit {
+                    // TODO decide how to treat this
+                }
```

Comment:
Do we need to error here? Can we just log it and keep going? Do we use `request.limit` anywhere else?

Reply:
I will start with making this an error so the batch logic is consistent, this should not happen since all requested headers fit in a single response.
But can phase that out with some additional checks